functional_programming_tests/test/without_functional_programming.cpp

```diff
@@ -0,0 +1,205 @@
+#include <gtest/gtest.h>
+
+#include <cmath>
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <vector>
+
+#include <rclcpp/rclcpp.hpp>
+#include <std_msgs/msg/u_int8_multi_array.hpp>
+
+using namespace std::chrono_literals;
+
+// production_code.h/cc
+template<typename T>
+using Map = std::vector<std::vector<T>>;
+
+// X, Y position
+struct Pos {
+	size_t x;
+	size_t y;
+};
+
+using Path = std::vector<Pos>;
+
+// Operator overload to print path
+std::ostream& operator<<(std::ostream& os, std::vector<Pos> const& path) {
+	for (auto const& pose:path) {
+		os << "(" << pose.x << ", " << pose.y << ")\n";
+	}
+	return os;
+}
+
+// Operator overload for position comparison
+bool operator==(Pos const& lhs, Pos const& rhs) {
+	return lhs.x == rhs.x && lhs.y == rhs.y;
+}
+
+class PathGenerator {
+public:
+	PathGenerator() : node_(std::make_shared<rclcpp::Node>("bad_test_subscriber")), executor_(std::make_shared<rclcpp::executors::SingleThreadedExecutor>()) {
+        // Create ROS2 subscriber for the costmap
+		costmap_subscriber_ = node_->create_subscription<std_msgs::msg::UInt8MultiArray>("/test/costmap", rclcpp::SystemDefaultsQoS(), [this](const std_msgs::msg::UInt8MultiArray::SharedPtr msg){costmap_callback(msg);});
+
+		// Let the node spin to execute get_costmap callbacks
+		executor_->add_node(node_);
+		executor_thread_ = std::thread([this]() { executor_->spin(); });    		
+	}
+	~PathGenerator() {
+		executor_->cancel();
+		executor_thread_.join();		
+	}
+    std::optional<Path> generate_global_path(Pos const& start, Pos const& goal);
+    void costmap_callback(const std_msgs::msg::UInt8MultiArray::SharedPtr msg);
+
+    Map<unsigned char> get_costmap() const { return map_;}
+private:
+	rclcpp::Node::SharedPtr node_;
+	rclcpp::Executor::SharedPtr executor_;
+	std::thread executor_thread_;
+
+	Map<unsigned char> map_;
+
+	rclcpp::Subscription<std_msgs::msg::UInt8MultiArray>::SharedPtr costmap_subscriber_;
+};
+
+// From the Robot, goal and costmap, generate a trajectory (Deterministic calculation)
+std::optional<Path> PathGenerator::generate_global_path(Pos const& start, Pos const& goal) { // Calculation
```

Comment:
Move the definitions into the class declaration since you're doing all of this in one file

Reply:
Oh, hey, I don't mean to make you work too hard, but this should only be accessible through a ros service call.
I mean it's bad enough that you need the costmap publisher but people's mindset is usually to have all the things only accessible through pub/sub, services, or actions. Because they instantiate usually one class per node and stuff like that (or that's how they start) so public methods wouldn't be needed.

Reply:
Should the service be something that is in another class then? Or should I also create the service in the PathGenerator class?

Reply:
In a conventional ros2 example, the service server would be a member variable of the class that also has the algorithm as a method. See the ros2 tutorials and extrapolate from those patterns https://docs.ros.org/en/foxy/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Publisher-And-Subscriber.html

Interesting, they didn't use a class here https://docs.ros.org/en/foxy/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Service-And-Client.html